### PR TITLE
Add a fuzzing target

### PR DESCRIPTION
**Pull request by ruuda**
This pull request adds a [cargo-fuzz](https://github.com/rust-fuzz/cargo-fuzz) fuzzing target. I added a readme that explains how to run the fuzzer. This requires a nightly toolchain, and it currently only works on x86-64 Linux and x86-64 Mac.

When seeded with a NEF file, it almost immediately finds an index out of bounds crash. The base64-encoded input that triggers this crash is `TU0AKgAAAAgAGQD+AAQAAA==`.

### CODE REVIEW

**Comment by pedrocr** [comment]
Thanks for this I've been meaning to do this for a while now. See my reddit comment for an explanation on why I don't consider that specific case a bug.

File: fuzz/fuzzers/decode.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#![no_main]
+#[macro_use] extern crate libfuzzer_sys;
+extern crate rawloader;
+
+use std::io;
+
+fuzz_target!(|data: &[u8]| {
```
Comment (pedrocr) [inline fuzz/fuzzers/decode.rs]: The official API is decode_safe() and that will catch panics. If we change this to that will hangs still be caught? Those are of course bugs.

Comment (ruuda) [inline fuzz/fuzzers/decode.rs]: `decode_safe()` takes a file, not an `io::Read`, so it cannot use the fuzzer-provided `&[u8]` without writing that to a file first. I wrapped the `decode()` in a a `panic::catch_unwind` instead.

**Comment by ruuda** [comment]
This will still detect crashes when Address Sanitizer reports an invalid memory access. But that is desired, right? Because if Rust did not prevent that, it can still be dangerous, even when catching panics.

Base64-encoded file that causes an Address Sanitizer crash:

```
TU0AKgAAAAgAGQD+AAQAAAABAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAEAAAQAAAABAAAAoAEBAAQAAAABAAAAeAECAAMAAAADAAABPAEDAAMAAAABAAEAAAEGAAMA
AAABAAIAAAEPAAIAAAASAAABRAEQAAIAAAAMAAABWAERAAQAAAABAAHo4AESAAMAAAABAAEAAAEV
AAMAAAABAAMAAAEWAAQAAAABAAAAeAEXAAQAAAABAADhAAEaAAUAAAABAAABZAEbAAUAAAABAAAB
bAEcAAMAAAABAAEAAAEoAANNTSoAAAAACAAAAACgAQEAAAABAAIAAAExAAIAAAAKAAABdAEyAAIA
AAAUAAABAIABSgAEAAAAAgAAAZQCFAAFAAAABgAAAZyHaQAEAAAAAQAAAeCIJQAEAAA6BAAAAAB4
AQIAAAEAAejOkAMAAgAAABQAAAHMkhYAAQAAAAQBAAAAAAAAAAAAAAgACAAIAABOSUtPTiBDT1JQ
T1JBVElPTgAAAE5JS09OIEQ1MDAwAAAAASwAAAABAAABLAAAAAFWZXIuMS4wMCAAAAAyMDE1OjA5
OjA1IDE1OjU1OjI4AAACyeAAAspYAAAAAAAAAAEAAAD/AAAAAQADAAAAAwAAATwBAwADAAAAAAAA
AAABAAAA/wAAAAEAAAAAAAAAAQAAAP8AAAABMjAxNTowOTowNSA=
```

**Comment by pedrocr** [comment]
Not sure how we can end up with an AS trigger. It's probably an out-of-bounds access that rust catches being caught by AS first, no? Can you run the file under the normal rawloader binary and see what it does?

File: fuzz/fuzzers/decode.rs (hunk 1 above)
Comment (pedrocr) [inline fuzz/fuzzers/decode.rs]: We need the &[u8] API that does the copying then.

**Comment by pedrocr** [comment]
The reason the normal program doesn't panic/crash is probably that because it adds zeroes at the end that call eventually finds one of those. Still don't know why that's an AS and not panic though.

**Comment by ruuda** [comment]
The normal binary does not crash, but that is no surprise; asan manipulates allocations to detect these errors more easily. The topmost stackframe reported is [`slice/mod.rs:1200`](https://github.com/rust-lang/rust/blob/272e77f035d8cca43f228764c2028c09eeaca20e/src/libcore/slice/mod.rs#L1200), which calls [`search_while`](https://github.com/rust-lang/rust/blob/272e77f035d8cca43f228764c2028c09eeaca20e/src/libcore/slice/mod.rs#L1261-L1278), but it looks ok to me. It does pointer comparisons in unsafe code, so there is nothing that would cause a panic there. It looks like `data` is always a slice of `buf`. Where does `buf` come from? Could it include invalid memory?

**Comment by pedrocr** [comment]
After some questions on #rust my best guess is that the synthetic file has an impossibly large tag that is somehow causing an invalid stack allocation somewhere. Those can indeed cause rust to SEGV. But maybe this is actually a bug in core rust?

**Comment by ruuda** [comment]
Does Rust even have runtime-sized stack allocations? The [alloca RFC](https://github.com/rust-lang/rfcs/pull/1808) was closed.

**Comment by pedrocr** [comment]
You can definitely blow the stack by allocating a lot but maybe that specific case I was mentioning is not possible. For the purpose of this PR I'd say we should disable AS if possible. At least for now I'd focus on cases that hang or crash the actual normal binary.

**Comment by ruuda** [comment]
Here’s a file that takes about a second to decode when compiled in release mode, even though it is only 1245 bytes:

```
TU0AKgAAAAgAGQAAAAAAAAD///////////////8AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB0dHR0dHR0dHR0d
HR0AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABNTQAqKgAAAAgAGQAA
AAAAAAD///////////////8AAAAAAAAAAAAAAAAAAAAAAP4ABAAAAAE6AAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAQFNTQAqAAAACAAZAP4ABAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAB0dHR0dHR0dHR0dHR0AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAABNTQAqKgAAAAgAGQAAAAAAAAD///////////////8AAAAAAAAAAHIABAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AToAAAABAU1NACoAAAAIABkA/gAEAAAAAQAAAAEBAAAEAAAAAAAAAQEAAAQAAAABAAAAoAEBOzoE
AACsAAB4AQIDAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACkAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAB0dHR0dHR0dHR0dHR0AAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAABNTQAqKgAAAAgAGQAAAAAAAAD///////////////8AAAAAAAAA
AAAAAAAAAAAAAP4ABAAAAAE6AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQFNTQAqAAAA
CAAZAP4ABAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB0dHR0dHR0d
HR0dHR0AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABNTQAqKgAAAAgA
GQAAAAAAAAD///////////////8AAAAAAAAAAAAAAAAAAAAAAP4ABAAAAAE6AAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAQFNTQAqAAAACAAZAP4ABAAAAAEAAAABAAEAAAQA
```

**Comment by pedrocr** [comment]
I've since added two fuzzing targets (one over metadata and another over the decoders) and "completed" several fuzzing sessions fixing a few bugs. On the way I also had to fix some bugs in afl-fuzz :) Thanks for pushing me to do this.

**Pull request closed without merging**